kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,1139 @@
+    }
+  }
+
+  test("Updating schema and tightening nullability on existing field fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", IntegerType.INTEGER, false, currentSchema.get("a").getMetadata)
+
+      assertSchemaEvolutionFails[IllegalArgumentException](
+        table,
+        engine,
+        newSchema,
+        "Cannot tighten the nullability of existing field a")
+    }
+  }
+
+  test("Cannot tighten nullability on array element") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+        .add(
+          "arr",
+          new ArrayType(StringType.STRING, true))
+
+      createTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", IntegerType.INTEGER, true, currentSchema.get("a").getMetadata)
+        .add(
+          "arr",
+          new ArrayType(StringType.STRING, false),
+          currentSchema.get("arr").getMetadata)
+
+      assertSchemaEvolutionFails[IllegalArgumentException](
+        table,
+        engine,
+        newSchema,
+        "Cannot tighten the nullability of existing field")
+    }
+  }
+
+  def fieldMetadataForColumn(
+      columnId: Long,
+      physicalColumnId: String): FieldMetadata = {
+    FieldMetadata.builder()
+      .putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, columnId)
+      .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalColumnId)
+      .build()
+  }
+
+  def fieldMetadataForArrayColumn(
+      columnId: Long,
+      physicalColumnId: String,
+      arrayFieldName: String,
+      nestedElementId: Long): FieldMetadata = {
+    FieldMetadata.builder()
+      .fromMetadata(fieldMetadataForColumn(columnId, physicalColumnId))
+      .putFieldMetadata(
+        ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY,
+        FieldMetadata.builder().putLong(s"$arrayFieldName.element", nestedElementId).build())
+      .build()
+  }
+
+  def fieldMetadataForMapColumn(
+      columnId: Long,
+      physicalColumnId: String,
+      mapFieldName: String,
+      keyId: Long,
+      valueId: Long): FieldMetadata = {
+    FieldMetadata.builder()
+      .fromMetadata(fieldMetadataForColumn(columnId, physicalColumnId))
+      .putFieldMetadata(
+        ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY,
+        FieldMetadata.builder().putLong(s"$mapFieldName.key", keyId)
+          .putLong(s"$mapFieldName.value", valueId).build())
+      .build()
+  }
+
+  private def createTable(
```

Comment:
we have one already in DeltaTableWriteSuiteBase

Reply:
Hm I don't see one? There's a few createTxn helpers but this was meant to wrap a lot of the same table setup stuff that we do throughout this test

Reply:
Ah saw your comment above